README.md
concurrency-limit
=================

[![PyPI](https://badge.fury.io/py/concurrency-limit.svg)](https://pypi.org/project/concurrency-limit/)
[![Test Status](https://github.com/anexia/python-concurrency-limit/actions/workflows/test.yml/badge.svg?branch=main)](https://github.com/anexia/python-concurrency-limit/actions/workflows/test.yml)
[![Codecov](https://codecov.io/gh/anexia/python-concurrency-limit/branch/main/graph/badge.svg)](https://codecov.io/gh/anexia/python-concurrency-limit)

`concurrency-limit` is a library that implements a distributed concurrency limiting mechanism using Redis as a
backend. The library allows to limit the number of concurrent executions of code sections, either by waiting until
the currently running execution threads go below the limit, or by raising an exception if there are currently too many
execution threads.

# Installation

With a [correctly configured](https://pipenv.pypa.io/en/latest/basics/#basic-usage-of-pipenv) `pipenv` toolchain:

```sh
pipenv install concurrency-limit
```

You may also use classic `pip` to install the package:

```sh
pip install concurrency-limit
```

# Getting started

## How it works
- A limit of concurrently running scopes of a concurrency group can be defined using the `limit` decorator. A 
  concurrency group is defined by the `key` attribute of the `LimitConfiguration` instance.
- If the count of concurrently running scopes of a concurrency group is below the configured limit, the scope is 
  executed immediately.
- If the count of concurrently running scopes of a concurrency group exceeds the configured limit, the context manager 
  wait unit it goes below the configured limit.
- If the count of concurrently running scopes of a concurrency group does not go below the configured limit with the 
  configured timeout, a `ConcurrencyLimitExceededException` exception is raised.

## Usage

### Example 1

Limit the concurrency group `"example-1"` to `100` concurrently running scopes. If there are already `100` running
scopes, wait until the count of concurrently running scopes go below `100`. Fail if this does not happen within `10` 
seconds by raising a `ConcurrencyLimitExceededException` exception.

```python
import concurrency_limit

redis_configuration = concurrency_limit.RedisConfiguration(
    host='127.0.0.1',
    port=6379,
)
limit_configuration = concurrency_limit.LimitConfiguration(
    key='example-1',
    limit=100,
    limit_timeout=10,
)

with concurrency_limit.limit(redis_configuration, limit_configuration):
    do_something_magic()
```

### Example 2

Limit the concurrency group `"example-1"` to `100` concurrently running scopes. If there are already `100` running
scopes, wait until the count of concurrently running scopes go below `100`. Fail if this does not happen within `10` 
seconds by raising a `ConcurrencyLimitExceededException` exception. Check if the concurrently running scopes 
are below the limit every `1` second.

```python
import concurrency_limit

redis_configuration = concurrency_limit.RedisConfiguration(
    host='127.0.0.1',
    port=6379,
)
limit_configuration = concurrency_limit.LimitConfiguration(
    key='example-2',
    limit=100,
    limit_timeout=0,
    limit_interval=1.0,
)

with concurrency_limit.limit(redis_configuration, limit_configuration):
    do_something_magic()
```

### Example 3

Limit the concurrency group `"example-3"` to `100` concurrently running scopes. If there are already `100` running
scopes, fail immediately by raising a `ConcurrencyLimitExceededException` exception.

```python
import concurrency_limit

redis_configuration = concurrency_limit.RedisConfiguration(
    host='127.0.0.1',
    port=6379,
)
limit_configuration = concurrency_limit.LimitConfiguration(
    key='example-3',
    limit=100,
    limit_timeout=0,
)

with concurrency_limit.limit(redis_configuration, limit_configuration):
    do_something_magic()
```

### Example 4

Limit the concurrency group `"example-4"` to `100` concurrently running scopes. The implementation of the 
concurrency group needs to now the number of the concurrently running scope.

```python
import concurrency_limit

redis_configuration = concurrency_limit.RedisConfiguration(
    host='127.0.0.1',
    port=6379,
)
limit_configuration = concurrency_limit.LimitConfiguration(
    key='example-4',
    limit=100,
)

with concurrency_limit.limit(redis_configuration, limit_configuration) as group_number:
    do_something_magic(group_number)
```

### Example 5

Limit the concurrency group `"example-5"` to `100` concurrently running scopes. There is already an existing connection
pool to Redis within your application, that `concurrency_limit` should use.

```python
import concurrency_limit

redis_configuration = concurrency_limit.RedisConfiguration(
    connection_pool=my_redis_connection_pool,
)
limit_configuration = concurrency_limit.LimitConfiguration(
    key='example-5',
    limit=100,
)

with concurrency_limit.limit(redis_configuration, limit_configuration):
    do_something_magic()
```

### Example 6

Limit the concurrency group `"example-6"` to `100` concurrently running scopes. A redis URL will be used instead of
separately specifying host and port for the redis connection.

```python
import concurrency_limit

redis_configuration = concurrency_limit.RedisConfiguration.from_url('redis://127.0.0.1:6379/0')
limit_configuration = concurrency_limit.LimitConfiguration(
    key='example-6',
    limit=100,
)

with concurrency_limit.limit(redis_configuration, limit_configuration):
    do_something_magic()
```

## Configuration options

### `RedisConfiguration`

#### `host: str`

Default: `None`

The hostname or IP of the Redis server.

#### `port: int`

Default: `6379`

The Redis server port.

#### `path: str`

Default: `None`

The socket path used if `unix_socket` is set.

#### `db: int`

Default: `0`

The Redis database to use.

#### `username: str`

Default: `None`

The authentication username.

#### `password: str`

Default: `None`

The authentication password.

#### `max_connections: int`

Default: `10`

The maximum connections on the connection pool for each process.

#### `timeout: int`

Default: `10`

The Redis server connection timeout.

#### `secure: bool`

Default: `False`

Use secure connection to Redis server.

#### `unix_socket: bool`

Default: `False`

Use UNIX socket connection to Redis server. Will ignore `secure`, if set as there is no SSL mode with sockets.

#### `connection_class: typing.Type[redis.Connection]`

Default: `None`

Redis connection class to use instead of `secure` and `unix_socket` fields, if set.

#### `connection_pool: redis.ConnectionPool`

Default: `None`

Use this connection pool instance instead of the other fields, if set. All other fields of the configuration
instance are ignored in this case.

### `LimitConfiguration`

#### `key: str`

The concurrency limit group identifier. Use the same key for different scopes that should use the same concurrency
counter. You may use different limit configurations for scopes that use the same key.

#### `limit: int`

The concurrency limit for the limited concurrency group (defined by the `key`). If there are more concurrent executions
than this limit allows, the execution may wait for the concurrency count to go below the limit, or may
fail by raising a `ConcurrencyLimitExceededException` exception.

#### `limit_timeout: int`

Default: `10`

The timeout that defines how long to wait for the concurrency count to go below the configured limit. The timeout
is configured in seconds. Set to `0` if you want to raise a `ConcurrencyLimitExceededException` exception immediately 
if there are too many concurrent executions.

#### `limit_interval: float`

Default: `0.1`

If there are too many concurrent executions of a scope, and a `limit_timeout` is set to a value greater than `0`, this
configuration defines the interval to re-check the current concurrency count. As soon as the concurrency count is 
below the configured limit, the execution of the scope starts.

#### `limit_expire: int`

Default: `60`

The expiry time of the concurrency count key, configured in seconds. If a concurrency count is untouched for the 
configured time, it will be deleted.

# Supported versions

|             | Supported |
|-------------|-----------|
| Python 3.9  | ✓         |
| Python 3.10 | ✓         |
| Python 3.11 | ✓         |
| Python 3.12 | ✓         |

# List of developers

* Andreas Stocker <AStocker@anexia.com>, Lead Developer
* Patrick Taibel <PTaibel@anexia.com>, Developer


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.1] - 2023-10-30
### Added
- Added support for Python 3.12

### Changed
- Handle WRONGTYPE errors from redis

## [1.1.0] - 2023-10-04
### Added
- Support for UNIX socket redis connections
- Generating redis configurations from a given URL

### Changed
- Minimum required version of `redis-py` is now `4.0`

### Removed
- Removed support for Python 3.7 and Python 3.8

### Fixed
- Individual concurrency locks expire despite connection errors

## [1.0.3] - 2022-01-19
### Changed
- Updated badges

## [1.0.2] - 2022-01-19
### Changed
- Updated repository URL

## [1.0.1] - 2021-11-13
### Fixed
- Fixed package name in README.md

## [1.0.0] - 2021-11-11
### Added
- Concurrency limit context-manager
- Support for creating and managing Redis connections
- Support for using existing Redis connection pools

[Unreleased]: https://github.com/anexia/python-concurrency-limit/compare/1.1.1...HEAD
[1.1.1]: https://github.com/anexia/python-concurrency-limit/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/anexia/python-concurrency-limit/compare/1.0.3...1.1.0
[1.0.3]: https://github.com/anexia/python-concurrency-limit/compare/1.0.2...1.0.3
[1.0.2]: https://github.com/anexia/python-concurrency-limit/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/anexia/python-concurrency-limit/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/anexia/python-concurrency-limit/releases/tag/1.0.0


tests/test_base.py
import collections
import fnmatch
import functools
import threading
import time

__all__ = ["RedisMock", "concurrent"]

import redis.exceptions


class RedisMock:
    def __init__(self):
        self._lock = threading.Lock()
        self._keys = collections.defaultdict(lambda: None)
        self._hashes = collections.defaultdict(lambda: {})
        self._expires = collections.defaultdict(lambda: time.time() + 2 ** 32)

    def scan_iter(self, match):
        keys = {*self._keys.keys(), *self._hashes.keys()}
        for key in fnmatch.filter(keys, match):
            yield key

    def set(self, name, value):
        with self._lock:
            self._keys[name] = value

    def delete(self, name):
        with self._lock:
            del self._keys[name]

    def hlen(self, name):
        with self._lock:
            self._ensure_type_hash(name)
            self._clean_expired(name)
            return len(self._hashes[name])

    def hset(self, name, key, value):
        with self._lock:
            self._ensure_type_hash(name)
            self._hashes[name][key] = str(value)

    def hdel(self, name, *keys):
        count = 0

        with self._lock:
            for key in keys:
                try:
                    self._ensure_type_hash(name)
                    del self._hashes[name][key]
                    count += 1
                except KeyError:
                    pass

        return count

    def hscan_iter(self, name):
        with self._lock:
            self._ensure_type_hash(name)
            self._clean_expired(name)
            _hash = dict(self._hashes[name])

        for hkey, hvalue in _hash.items():
            yield hkey, hvalue

    def expire(self, name, _time):
        with self._lock:
            self._expires[name] = _time + time.time()

    def pipeline(self):
        client = self

        class _Pipeline:
            def __init__(self):
                self.buffer = []

            def __getattr__(self, item):
                def _wrapper(*args, **kwargs):
                    self.buffer.append(wrapped(*args, **kwargs))
                    return self

                wrapped = getattr(client, item)

                return _wrapper

            def execute(self):
                try:
                    return self.buffer
                finally:
                    self.buffer = []

        return _Pipeline()

    def _ensure_type_hash(self, name):
        if name in self._keys:
            raise redis.ResponseError(
                "WRONGTYPE Operation against a key holding the wrong kind of value"
            )

    def _clean_expired(self, name):
        if time.time() > self._expires[name]:
            del self._expires[name]

            try:
                del self._hashes[name]
            except KeyError:
                pass


def concurrent(threads: int):
    class ExceptionAwareThread(threading.Thread):
        def run(self):
            try:
                setattr(self, "_ret", self._target(*self._args, **self._kwargs))
            except BaseException as e:
                setattr(self, "_exc", e)

        def join(self, timeout=None):
            super().join(timeout=timeout)

            exc = getattr(self, "_exc", None)
            ret = getattr(self, "_ret", None)

            if exc:
                raise exc
            else:
                return ret

    def _decorator(func):
        @functools.wraps(func)
        def _wrapper(*args, **kwargs):
            ts = []

            # create the threads
            for i in range(threads):
                ts.append(ExceptionAwareThread(target=func, args=args, kwargs=kwargs))

            # start the threads
            for t in ts:
                t.start()

            # wait for all threads to finish
            for t in ts:
                t.join()

        return _wrapper

    return _decorator


tests/test_limit.py
import time

import pytest
import pytest_mock

import concurrency_limit

from test_base import *


def test_limit_without_concurrency(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    with concurrency_limit.limit(
        concurrency_limit.RedisConfiguration(),
        concurrency_limit.LimitConfiguration(key="key-1", limit=1),
    ) as slot_id:
        assert slot_id == 1


def test_limit_slot_ids(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    slot_ids = []

    @concurrent(threads=10)
    def _concurrent_function():
        with concurrency_limit.limit(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(
                key="key-1", limit=10, limit_timeout=0
            ),
        ) as slot_id:
            slot_ids.append(slot_id)
            time.sleep(1)

    _concurrent_function()

    slot_ids.sort()
    assert slot_ids == list(range(1, 11))


def test_limit_within_limit(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    @concurrent(threads=1)
    def _concurrent_function():
        with concurrency_limit.limit(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(key="key-1", limit=1, limit_timeout=0),
        ):
            time.sleep(1)

    _concurrent_function()


def test_limit_exceeded_limit_without_timeout(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    @concurrent(threads=10)
    def _concurrent_function():
        with concurrency_limit.limit(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(key="key-1", limit=1, limit_timeout=0),
        ):
            time.sleep(1)

    with pytest.raises(concurrency_limit.ConcurrencyLimitException):
        _concurrent_function()


def test_limit_exceeded_limit_within_timeout(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    @concurrent(threads=10)
    def _concurrent_function():
        with concurrency_limit.limit(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(key="key-1", limit=5, limit_timeout=5),
        ):
            time.sleep(1)

    _concurrent_function()


def test_limit_exceeded_limit_exceeded_timeout(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    @concurrent(threads=10)
    def _concurrent_function():
        with concurrency_limit.limit(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(key="key-1", limit=5, limit_timeout=5),
        ):
            time.sleep(10)

    with pytest.raises(concurrency_limit.ConcurrencyLimitException):
        _concurrent_function()


def test_limit_within_limit_expire(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    with concurrency_limit.limit(
        concurrency_limit.RedisConfiguration(),
        concurrency_limit.LimitConfiguration(
            key="key-1", limit=1, limit_expire=5, limit_timeout=0
        ),
    ):
        time.sleep(1)

        with pytest.raises(concurrency_limit.ConcurrencyLimitException):
            with concurrency_limit.limit(
                concurrency_limit.RedisConfiguration(),
                concurrency_limit.LimitConfiguration(
                    key="key-1", limit=1, limit_expire=5, limit_timeout=0
                ),
            ):
                time.sleep(1)


def test_limit_exceeded_limit_expire(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    with concurrency_limit.limit(
        concurrency_limit.RedisConfiguration(),
        concurrency_limit.LimitConfiguration(
            key="key-1", limit=1, limit_expire=5, limit_timeout=0
        ),
    ):
        time.sleep(10)

        with concurrency_limit.limit(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(
                key="key-1", limit=1, limit_expire=5, limit_timeout=0
            ),
        ):
            time.sleep(1)


def test_limit_with_high_load(mocker: pytest_mock.MockerFixture):
    mocker.patch(
        "concurrency_limit.context_managers.get_redis", return_value=RedisMock()
    )

    counter = 0

    @concurrent(threads=10_000)
    def _concurrent_function():
        with concurrency_limit.limit(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(
                key="key-1", limit=500, limit_timeout=1
            ),
        ) as slot_id:
            nonlocal counter

            counter += 1

            assert 1 <= slot_id <= 500
            assert 1 <= counter <= 500

            time.sleep(1.5)
            counter -= 1

    with pytest.raises(concurrency_limit.ConcurrencyLimitException):
        _concurrent_function()


def test_limit_wrong_type(mocker: pytest_mock.MockerFixture):
    client = RedisMock()
    mocker.patch("concurrency_limit.context_managers.get_redis", return_value=client)

    client.set("key-1", int(time.time()) + 10)

    with concurrency_limit.limit(
        concurrency_limit.RedisConfiguration(),
        concurrency_limit.LimitConfiguration(key="key-1", limit=1, limit_timeout=5),
    ):
        time.sleep(1)


def test_limit_clean(mocker: pytest_mock.MockerFixture):
    client = RedisMock()
    mocker.patch("concurrency_limit.utils.get_redis", return_value=client)

    client.hset("key-1", "expired-1", int(time.time()) - 10)
    client.hset("key-1", "expired-2", int(time.time()) - 10)
    client.hset("key-1", "unexpired-1", int(time.time()) + 10)
    client.hset("key-1", "unexpired-2", int(time.time()) + 10)

    assert client.hlen("key-1") == 4
    assert (
        concurrency_limit.limit_clean(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(key="key-1", limit=1),
        )
        == 2
    )
    assert client.hlen("key-1") == 2

    for scan_key in concurrency_limit.limit_iter(
        concurrency_limit.RedisConfiguration(), "key-*"
    ):
        for scan_lock_id, _ in client.hscan_iter(scan_key):
            assert scan_lock_id.startswith("unexpired-")


def test_limit_clean_expire_wrong_type(mocker: pytest_mock.MockerFixture):
    client = RedisMock()
    mocker.patch("concurrency_limit.utils.get_redis", return_value=client)

    client.hset("key-1", "valid-1", int(time.time()) + 10)
    client.hset("key-1", "valid-2", int(time.time()) + 10)
    client.hset("key-1", "invalid-1", "abc")
    client.hset("key-1", "invalid-2", "def")

    assert client.hlen("key-1") == 4
    assert (
        concurrency_limit.limit_clean(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(key="key-1", limit=1),
        )
        == 2
    )
    assert client.hlen("key-1") == 2

    for scan_key in concurrency_limit.limit_iter(
        concurrency_limit.RedisConfiguration(), "key-*"
    ):
        for scan_lock_id, _ in client.hscan_iter(scan_key):
            assert scan_lock_id.startswith("valid-")


def test_limit_clean_wrong_type(mocker: pytest_mock.MockerFixture):
    client = RedisMock()
    mocker.patch("concurrency_limit.utils.get_redis", return_value=client)

    client.set("key-1", int(time.time()) + 10)

    assert (
        concurrency_limit.limit_clean(
            concurrency_limit.RedisConfiguration(),
            concurrency_limit.LimitConfiguration(key="key-1", limit=1),
        )
        == 1
    )


tests/test_configuration.py
import typing

import redis.connection

import pytest

from concurrency_limit import RedisConfiguration


@pytest.mark.parametrize(
    "url,config",
    [
        (
            "redis://127.0.0.1:6379/1",
            RedisConfiguration(host="127.0.0.1", port=6379, db=1),
        ),
        (
            "redis://test-user:test-pw@127.0.0.1:6379/1",
            RedisConfiguration(
                host="127.0.0.1",
                port=6379,
                db=1,
                username="test-user",
                password="test-pw",
            ),
        ),
        (
            "rediss://127.0.0.1:6379/1",
            RedisConfiguration(
                host="127.0.0.1", port=6379, db=1, connection_class=redis.SSLConnection
            ),
        ),
        (
            "rediss://test-user:test-pw@127.0.0.1:6379/1",
            RedisConfiguration(
                host="127.0.0.1",
                port=6379,
                db=1,
                username="test-user",
                password="test-pw",
                connection_class=redis.SSLConnection,
            ),
        ),
        (
            "unix:///run/redis.sock?db=1",
            RedisConfiguration(
                path="/run/redis.sock",
                port=6379,
                db=1,
                connection_class=redis.UnixDomainSocketConnection,
            ),
        ),
        (
            "unix://test-user:test-pw@/run/redis.sock?db=1",
            RedisConfiguration(
                path="/run/redis.sock",
                port=6379,
                db=1,
                username="test-user",
                password="test-pw",
                connection_class=redis.UnixDomainSocketConnection,
            ),
        ),
    ],
)
def test_configuration_from_url(url: str, config: RedisConfiguration):
    assert RedisConfiguration.from_url(url) == config


@pytest.mark.parametrize(
    "url,kwargs,config",
    [
        (
            "redis://127.0.0.1:6379/1",
            {"connection_class": redis.SSLConnection},
            RedisConfiguration(
                host="127.0.0.1", port=6379, db=1, connection_class=redis.SSLConnection
            ),
        ),
        (
            "redis://127.0.0.1:6379/1",
            {"connection_class": redis.UnixDomainSocketConnection},
            RedisConfiguration(
                host="127.0.0.1",
                port=6379,
                db=1,
                connection_class=redis.UnixDomainSocketConnection,
            ),
        ),
    ],
)
def test_configuration_from_url_with_kwargs(
    url: str, kwargs: dict, config: RedisConfiguration
):
    assert RedisConfiguration.from_url(url, **kwargs) == config


@pytest.mark.parametrize(
    "config,expected_class",
    [
        (RedisConfiguration(), redis.Connection),
        (RedisConfiguration(secure=True), redis.SSLConnection),
        (RedisConfiguration(unix_socket=True), redis.UnixDomainSocketConnection),
        (RedisConfiguration(connection_class=redis.Connection), redis.Connection),
        (RedisConfiguration(connection_class=redis.SSLConnection), redis.SSLConnection),
        (
            RedisConfiguration(connection_class=redis.UnixDomainSocketConnection),
            redis.UnixDomainSocketConnection,
        ),
        (
            RedisConfiguration(secure=True, unix_socket=True),
            redis.UnixDomainSocketConnection,
        ),
        (
            RedisConfiguration(
                secure=True, unix_socket=True, connection_class=redis.Connection
            ),
            redis.Connection,
        ),
    ],
)
def test_configuration_get_connection_class(
    config: RedisConfiguration, expected_class: typing.Type[redis.Connection]
):
    assert config.get_connection_class() == expected_class


concurrency_limit/utils.py
import time

import redis

from ._connections import *
from .configuration import *

__all__ = ["limit_clean", "limit_iter"]


def limit_clean(
    redis_configuration: RedisConfiguration, limit_configuration: LimitConfiguration
):
    """
    Cleans stale limit locks in the hash for the given limit configuration.

    :param redis_configuration: RedisConfiguration object containing the configuration details for connecting to Redis.
    :param limit_configuration: LimitConfiguration object containing the configuration details for the limit.
    :return: Number of cleaned items
    """
    client = get_redis(redis_configuration)
    current = int(time.time())
    count = 0

    lock_key = limit_configuration.key

    # If the key does not contain a hash, Redis fails with a WRONGTYPE exception that
    # we handle by deleting the key and re-trying.
    try:
        for scan_lock_id, scan_lock_expire in client.hscan_iter(lock_key):
            try:
                clean_lock = current >= int(scan_lock_expire)
            except (ValueError, TypeError):
                clean_lock = True

            if clean_lock:
                count += client.hdel(lock_key, scan_lock_id)

    except redis.ResponseError as exc:
        if str(exc).startswith("WRONGTYPE"):
            client.delete(lock_key)
            return 1

        raise  # pragma: no cover

    return count


def limit_iter(redis_configuration: RedisConfiguration, key_pattern: str):
    """
    Return an iterator over the items in Redis specified by `key_pattern`
    using the Redis connection specified by `redis_configuration`.

    :param redis_configuration: The configuration for connecting to Redis.
    :param key_pattern: The pattern for Redis to iterate over.
    :return: Scan Iterator
    """
    return get_redis(redis_configuration).scan_iter(key_pattern)


concurrency_limit/_connections.py
import functools
import threading

import redis

from .configuration import *

__all__ = ["get_redis"]

_connection_pool_map = {}
_connection_pool_lock = threading.Lock()


@functools.cache
def get_redis(configuration: RedisConfiguration) -> redis.Redis:
    """
    Gets the Redis client used to store the concurrency keys.

    :param configuration: Redis connection configuration
    :return: Redis client
    """
    if configuration.connection_pool:
        return _get_redis_by_connection_pool(configuration)

    else:
        return _get_redis_by_credentials(configuration)


def _get_redis_by_credentials(configuration: RedisConfiguration) -> redis.Redis:
    """
    Gets the Redis client used to store the concurrency keys using the connection credentials on the configuration
    object. This method creates a connection pool using the credentials, als will use this connection pool for
    all subsequent calls of this method with the same configuration.

    :param configuration: Redis connection configuration
    :return: Redis client
    """
    global _connection_pool_map
    global _connection_pool_lock

    with _connection_pool_lock:
        if configuration not in _connection_pool_map:
            _connection_pool_map[configuration] = redis.BlockingConnectionPool(
                host=configuration.host,
                port=configuration.port,
                path=configuration.path,
                db=configuration.db,
                username=configuration.username,
                password=configuration.password,
                max_connections=configuration.max_connections,
                timeout=configuration.timeout,
                connection_class=configuration.get_connection_class(),
            )

        return redis.Redis(connection_pool=_connection_pool_map[configuration])


def _get_redis_by_connection_pool(configuration: RedisConfiguration) -> redis.Redis:
    """
    Gets the Redis client used to store the concurrency keys using the connection pool on the configuration
    object.

    :param configuration: Redis connection configuration
    :return: Redis client
    """
    return redis.Redis(connection_pool=configuration.connection_pool)


concurrency_limit/configuration.py
import dataclasses
import typing

import redis
import redis.connection

__all__ = ["RedisConfiguration", "LimitConfiguration"]


@dataclasses.dataclass(eq=True, frozen=True)
class RedisConfiguration:
    """
    Redis connection configuration.
    """

    host: str = None
    "The hostname or IP of the Redis server."

    port: int = 6379
    "The Redis server port."

    path: str = None
    "The socket path used if `unix_socket` is set."

    db: int = 0
    "The Redis database to use."

    username: str = None
    "The authentication username."

    password: str = None
    "The authentication password."

    max_connections: int = 10
    "The maximum connections on the connection pool."

    timeout: int = 10
    "The Redis server connection timeout."

    secure: bool = False
    "Use secure connection to Redis server."

    unix_socket: bool = False
    "Use UNIX socket connection to Redis server. Will ignore `secure`, if set."

    connection_class: typing.Type[redis.connection.AbstractConnection] = None
    "Redis connection class to use instead of `secure` and `unix_socket` fields, if set."

    connection_pool: redis.ConnectionPool = None
    "Use this connection pool instance instead of the other fields, if set."

    def get_connection_class(self) -> typing.Type[redis.connection.AbstractConnection]:
        """
        Returns the `redis.Connection` class to use based on this configuration.

        Priority of configuration values:

        1. `connection_class` if set
        2. `redis.UnixDomainSocketConnection` if `unix_socket` is set
        3. `redis.SSLConnection` if `secure` is set
        4. `redis.Connection` as last resort

        :return: Determined `redis.Connection` class
        """
        if self.connection_class is not None:
            return self.connection_class

        if self.unix_socket:
            return redis.UnixDomainSocketConnection

        if self.secure:
            return redis.SSLConnection

        return redis.Connection

    @classmethod
    def from_url(cls, url, **kwargs):
        """
        Returns a `RedisConfiguration` object based on the given URL. This behaves exactly the same as the `from_url`
        methods of the `redis-py` package.

        Examples:

        - redis://[[username]:[password]]@localhost:6379/0
        - rediss://[[username]:[password]]@localhost:6379/0
        - unix://[[username]:[password]]@/path/to/socket.sock?db=0

        :param url: Redis URL to generate a `RedisConfiguration` from
        :param kwargs: Additional arguments to set for the configuration object
        """
        url_options = redis.connection.parse_url(url)

        if "connection_class" in kwargs:
            url_options["connection_class"] = kwargs["connection_class"]

        kwargs.update(url_options)
        return cls(**kwargs)


@dataclasses.dataclass(eq=True, frozen=True)
class LimitConfiguration:
    """
    Concurrency limit configuration.
    """

    key: str
    "The concurrency limit group identifier."

    limit: int
    "The limit for concurrently running executions of a concurrency group."

    limit_timeout: int = 10
    "Wait time for acquiring an execution slot before raising a `ConcurrencyLimitExceededException` exception."

    limit_interval: float = 0.1
    "Wait time between attempts to acquire an execution slot."

    limit_expire: int = 60
    "Expire time for the concurrency counter on Redis."
